staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go

```diff
@@ -0,0 +1,15 @@
+//go:build !run_integration_tests && !skip_special_tests
+// +build !run_integration_tests,!skip_special_tests
+
+package conversion
+
+import "testing"
+
+func TestFailRequireTags(t *testing.T) {
```

Comment:
One problem is that `go test -run=TestWebhookConverterWithWatchCache ./staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion` will not trigger this.

Perhaps we could put a be build error in this file?

There could be a line with just literally the warning text - I haven't tried that, though.

Reply:
good point.  How's this?

```
$ git diff
diff --git a/staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go b/staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go
index 3c4a03ed60a..22918d969fa 100644
--- a/staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go
+++ b/staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go
@@ -20,10 +20,15 @@ package conversion
 
 import "testing"
 
-func TestFailRequireTags(t *testing.T) {
-       t.Fatalf(`
-This directory contains tests which require special handling.
-If you intended to run these tests, use "-tags=run_integration_tests", otherwise use
-"-tags=skip_special_tests".
-`)
-}
+// NOTE: this intentionally fails to compile.
+`
+
+
+  !!!!!
+  This directory contains tests which require special handling.
+  If you intended to run these tests, use "-tags=run_integration_tests", otherwise use
+  "-tags=skip_special_tests".
+  !!!!!
+
+
+`
```

and:

```
$ go test -run=TestWebhookConverterWithWatchCache ./staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion
# k8s.io/apiextensions-apiserver/test/integration/conversion
staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion/special_test_tag_test.go:24:1: expected declaration, found `


  !!!!!
  This directory contains tests which require special handling.
  If you intended to run these tests, use "-tags=run_integration_tests", otherwise use
  "-tags=skip_special_tests".
  !!!!!


`
FAIL	k8s.io/apiextensions-apiserver/test/integration/conversion [setup failed]
FAIL
```

Reply:
Ugh.  So so so many things depend on compileable code - from gofmt to verify-import-aliases - all of them would have to be touched to either treat these files as special or learn how to handle build tags.  I think this is no good, but your problem still exists.

Reply:
How about: 

```
//go:build !run_e2e_tests && !skip_special_tests

package main

var _ = ____
var _ = ____This_directory_contains_tests_which_require_special_handling
var _ = ____If_you_intended_to_run_these_tests_use_the_tag
var _ = ________run_integration_tests
var _ = ____otherwise_use_the_tag
var _ = ________skip_special_tests
var _ = ____
```

So:

```
$ go test ./test/e2e_node/plugins/gcp-credential-provider
# k8s.io/kubernetes/test/e2e_node/plugins/gcp-credential-provider [k8s.io/kubernetes/test/e2e_node/plugins/gcp-credential-provider.test]
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:21:9: undefined: ____
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:22:9: undefined: ____This_directory_contains_tests_which_require_special_handling
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:23:9: undefined: ____If_you_intended_to_run_these_tests_use_the_tag
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:24:9: undefined: ________run_integration_tests
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:25:9: undefined: ____otherwise_use_the_tag
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:26:9: undefined: ________skip_special_tests
test/e2e_node/plugins/gcp-credential-provider/special_test_tag_test.go:27:9: undefined: ____
FAIL	k8s.io/kubernetes/test/e2e_node/plugins/gcp-credential-provider [build failed]
FAIL
```

But it's valid go.  Can you think of a more clever way to force an error?  OR maybe trick Go into using a unicode character that looks more like whitespace?

Reply:
> So so so many things depend on compileable code

Yes, that's a valid concern.

> Can you think of a more clever way to force an error?

No.

> OR maybe trick Go into using a unicode character that looks more like whitespace?

https://stackoverflow.com/questions/33947871/does-go-golang-actually-permit-unicode-characters-in-variable-names has some nice links to what are valid characters in an identifier. They are letters and digits in various incarnations, but nothing that looks like a whitespace - probably not a coincidence!

Underscore aren't too bad, but perhaps we should avoid using them in the tags themselves?

run_integration_tests -> runIntegrationTests
skip_special_tests -> skipSpecialTests

Reply:
Apropos `test/e2e_node/plugins/gcp-credential-provider/main_test.go`: that looks like a normal unit test to me.

In general, everything under test/e2e* is a normal unit test except for the files which contain the `ginkgo.RunSpecs` call:
```
test/e2e/e2e.go:        ginkgo.RunSpecs(t, "Kubernetes e2e suite", suiteConfig, reporterConfig)
test/e2e_kubeadm/e2e_kubeadm_suite_test.go:     ginkgo.RunSpecs(t, "E2EKubeadm suite", suiteConfig, reporterConfig)
test/e2e_node/e2e_node_suite_test.go:   ginkgo.RunSpecs(t, "E2eNode Suite", suiteConfig, reporterConfig)
```

Reply:
The problem with the Ginkgo suites is that `make WHAT=test/e2e/e2e.test` fails to build the e2e.test binaries. It must set the necessary build tag... somewhere. Not sure where.

Reply:
Yeah, I'm seeing this too -- if it is a compile error, then we need to hack on build.sh in addition to test.sh.  I don't love that.

Will think some more.

Reply:
```
+import (
+       "fmt"
+       "os"
+)
+
+func init() {
+       fmt.Print(`
+    This directory contains tests which require special handling.
+    If you intended to run these tests, use "-tags=run_kubeadm_cmd_tests", otherwise use
+    "-tags=skip_special_tests".
+
+`)
+       os.Exit(1)
+}
```

?

Reply:
The init looks promising. Will try it out on Monday.

Reply:
The problem with the Ginkgo suites is that `make WHAT=test/e2e/e2e.test` fails to build the e2e.test binaries.

We could NOT do that...?  ginkgo can take a normal go package target, e.g. `./test/e2e/...` and it will DTRT, as long as you pass the right tags.

This whole approach (tags) illustrates a significant diff from using regex - the tests are actually not compiled in, vs. being selected at runtime.

Reply:
I'm not sure I follow. `make WHAT=test/e2e/e2e.test` is something we want, because too many jobs depend on it and the resulting `e2e.test` is part of our binary release artifacts. It's what people download and run when checking a cluster for conformance.

In the case of `e2e_node` we need the binary because it needs to be copied into remote machines to run it alongside a container runtime and kubelet.

Were you suggesting that people should do `ginkgo run ./test/e2e` with the right tags instead?

Reply:
Yes.  I am saying it's fine if we want to build the e2e binary for the test, but having ourt own scripts assuming it is built makes CI damn near impossible to change.

Reply:
ugh, even trying to do this, it fails.  We do things with flags to TestMain (e.g. hack/verify-e2e-images.sh) that I just cannot figure out how to make work with `go test`.

Reply:
I'm not sure whether it's realistic to drop `make WHAT=test/e2e/e2e.test`. I suspect it's assumed to be working in too many different places.

"trying to do this" - what? I can help with the flags and `hack/verify-e2e-images.sh`, but I need to know what you are trying that isn't working.

https://github.com/kubernetes/kubernetes/blob/af2bf2d8f1b3437fe67e1686266a176320422465/hack/verify-e2e-images.sh#L37

calls the e2e.test binary. The corresponding `go test` invocation should be (untested):
```
go test -v ./test/e2e -args --list-images
```

Reply:
"this" was "don't pre-build the test binary".  I am abandoning that idea.  

I assumed what you saw d, but it doesn't work.  It depends on all the other test plugs being linked, to think?

This whole approach is starting to feel bad, like it might be correct but we have too much built up around the current patterns.

Reply:
`go test -v ./test/e2e -args --list-images` indeed depends on "everything" compiled into the binary.

Reply:
UGH!  I was missing `-v` !!!

Reply:
> I'm not sure whether it's realistic to drop make WHAT=test/e2e/e2e.test. I suspect it's assumed to be working in too many different places.

Yeah.